symsrv-fetch.py

```diff
@@ -24,345 +24,468 @@
+
+    log.info(f"Collected {len(to_dump)} files to dump")
+
+    return to_dump, stats
+
+
+async def fetch_and_write(output, client, filename, file_id):
+    path = os.path.join(filename, file_id, filename)
+    data = await fetch_file(client, MICROSOFT_SYMBOL_SERVER, path)
+
+    if not data:
+        return False
 
-    # Write an index file
-    with zipfile.ZipFile(args.zip, 'w', zipfile.ZIP_DEFLATED) as z:
+    output_dir = os.path.join(output, filename, file_id)
+    if not os.path.exists(output_dir):
+        os.makedirs(output_dir)
+
+    output_path = os.path.join(output_dir, filename)
+    async with AIOFile(output_path, "wb") as Out:
+        await Out.write(data)
+
+    return True
+
+
+async def fetch_all_helper(output, modules):
+    loop = asyncio.get_event_loop()
+    tasks = []
+    connector = TCPConnector(limit=100, limit_per_host=20)
+    async with ClientSession(
+        loop=loop, timeout=ClientTimeout(total=TIMEOUT), connector=connector
+    ) as client:
+        for filename, debug_id, code_file, code_id, has_code in modules:
+            tasks.append(fetch_and_write(output, client, filename, debug_id))
+            if has_code:
+                tasks.append(fetch_and_write(output, client, code_file, code_id))
+
+        return await asyncio.gather(*tasks)
+
+
+def fetch_all(output, modules):
+    res = asyncio.run(fetch_all_helper(output, modules))
+    res = iter(res)
+    fetched_modules = []
+    for filename, debug_id, code_file, code_id, has_code in modules:
+        fetched_pdb = next(res)
+        if has_code:
+            has_code = next(res)
+        if fetched_pdb:
+            fetched_modules.append((filename, debug_id, code_file, code_id, has_code))
+    return fetched_modules
+
+
+def get_base_data(url):
+    async def helper(url):
+        return await asyncio.gather(
+            fetch_missing_symbols(url),
+            # Symbols that we know belong to us, so don't ask Microsoft for them.
+            get_list("blacklist.txt", "Blacklist"),
+            # Symbols that we know belong to Microsoft, so don't skiplist them.
+            get_list("known-microsoft-symbols.txt", "Known Microsoft symbols"),
+            # Symbols that we've asked for in the past unsuccessfully
+            get_skiplist(),
+        )
+
+    return asyncio.run(helper(url))
+
+
+def gen_zip(output, output_dir, file_index):
+    if not file_index:
+        return
+
+    with zipfile.ZipFile(output, "w", zipfile.ZIP_DEFLATED) as z:
         for f in file_index:
-            z.write(os.path.join(symbol_path, f), f)
-    log.info('Wrote zip as %s' % args.zip)
+            z.write(os.path.join(output_dir, f), f)
+    log.info(f"Wrote zip as {output}")
+
+
+def retry_run(cmd, *arg):
+    for _ in range(5):
+        try:
+            return cmd(*arg)
+        except OSError as e:
+            # Sometimes for any reasons we've "Too much open files"
+            # So wait and try again
```

Comment:
Mmmh, this is not great, but perhaps due to the large number of handles.
Maybe instead of doing this and waste the work done when we get the error, we can just increase the limit with `ulimit -n`.

Reply:
(In the shell file which executes this)

Reply:
I almost fixed it in using TCPConnector(limit=100, limit_per_host=20) when creating the client (limit_per_host option): it's two times slower but it seems to work.
And what is strange here, is that it's an intermittent error (I wonder if it's related to MS server).
I hope we won't never go there but I let it to be sure that everything will run.

Reply:
@calixteman we can just use ulimit in the shell to increase the number of handles and completely avoid this

Reply:
I removed this stuff since it appears that TC has a max open files == 1048576.